### PR TITLE
adds functions to get data from facebook graph api

### PR DESCRIPTION
**Pull request by sherali-arbisoft**
adds 5 functions to get data of page and page posts with multiple metric insights from facebook graph api

### CODE REVIEW

File: digital_benchmark/digital_benchmark/settings.py [hunk 1]
```diff
@@ -37,6 +37,7 @@
     'django.contrib.sessions',
     'django.contrib.messages',
     'django.contrib.staticfiles',
+    'facebook_benchmark.apps.FacebookBenchmarkConfig', #register facebook_benchmark app
```
Comment (ehmadzubair) [inline digital_benchmark/digital_benchmark/settings.py]: only `facebook_benchmark` will be enough

File: digital_benchmark/digital_benchmark/settings.py [hunk 2]
```diff
@@ -118,3 +119,11 @@
 # https://docs.djangoproject.com/en/2.2/howto/static-files/
 
 STATIC_URL = '/static/'
+
+# facebook_benchmark settings
+facebook_graph_api_version = '3.1'
```
Comment (ehmadzubair) [inline digital_benchmark/digital_benchmark/settings.py]: use ALL_CAPS convention for constants (matching the rest of the constant names)

File: digital_benchmark/digital_benchmark/settings.py [hunk 3]
```diff
@@ -118,3 +119,11 @@
...
+facebook_default_fields_for_page = 'id,displayed_message_response_time,engagement{count},fan_count,name,overall_star_rating,rating_count,talking_about_count,unread_message_count,unread_notif_count,unseen_message_count,verification_status'
```
Comment (ehmadzubair) [inline digital_benchmark/digital_benchmark/settings.py]: Can these be in an array?

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+.vscode/
```
Comment (ehmadzubair) [inline .gitignore]: add .pyc files in gitignore

File: digital_benchmark/digital_benchmark/settings.py (hunk 2 above)
Comment (sherali-arbisoft) [reply digital_benchmark/digital_benchmark/settings.py]: ok

File: digital_benchmark/facebook_benchmark/data_provider.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import facebook as fb
+import requests
+
+from digital_benchmark.settings import facebook_graph_api_version, facebook_default_fields_for_page, facebook_default_fields_for_feed, facebook_default_fields_for_post, facebook_default_metrices_for_page_insights, facebook_default_metrices_for_post_insights
+
+class FacebookDataProvider:
+    def __init__(self, page_access_token, *args, **kwargs):
+        self.page_access_token = page_access_token
+        self.graph = fb.GraphAPI(access_token=page_access_token, version=facebook_graph_api_version)
```
Comment (ehmadzubair) [inline digital_benchmark/facebook_benchmark/data_provider.py]: `self.graph_api_client`

File: digital_benchmark/facebook_benchmark/data_provider.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import facebook as fb
+import requests
+
+from digital_benchmark.settings import facebook_graph_api_version, facebook_default_fields_for_page, facebook_default_fields_for_feed, facebook_default_fields_for_post, facebook_default_metrices_for_page_insights, facebook_default_metrices_for_post_insights
```
Comment (ehmadzubair) [inline digital_benchmark/facebook_benchmark/data_provider.py]: `from django.conf import settings`

File: digital_benchmark/digital_benchmark/settings.py (hunk 3 above)
Comment (sherali-arbisoft) [reply digital_benchmark/digital_benchmark/settings.py]: we have to send them as comma separated string to graph api
if we use them as array/list them we can join them with string join function

File: digital_benchmark/facebook_benchmark/data_provider.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            fields = facebook_default_fields_for_page
+        page = self.graph.get_object(id='me', fields=fields)
+        return page
+    
+    def get_all_posts(self, fields=''):
+        if not fields:
+            fields = facebook_default_fields_for_feed
+        feed = self.graph.get_connections(id='me', connection_name='feed', fields=fields)
+        all_posts = feed['data']
+        while 'paging' in feed and 'next' in feed['paging']:
```
Comment (ehmadzubair) [inline digital_benchmark/facebook_benchmark/data_provider.py]: 👍

File: digital_benchmark/facebook_benchmark/data_provider.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        self.graph = fb.GraphAPI(access_token=page_access_token, version=facebook_graph_api_version)
+    
+    def get_page_details(self, fields=''):
+        if not fields:
+            fields = facebook_default_fields_for_page
+        page = self.graph.get_object(id='me', fields=fields)
+        return page
+    
+    def get_all_posts(self, fields=''):
+        if not fields:
```
Comment (ehmadzubair) [inline digital_benchmark/facebook_benchmark/data_provider.py]: try not to change the value of the passed argument inside the function. There might be a use case in the future when we might need it. Make a new var

File: digital_benchmark/facebook_benchmark/data_provider.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        all_posts = feed['data']
+        while 'paging' in feed and 'next' in feed['paging']:
+            feed = requests.get(feed['paging']['next']).json()
+            if feed['data']:
+                all_posts += feed['data']
+        return all_posts
+    
+    def get_post_details(self, post_id, fields=''):
+        if not fields:
+            fields = facebook_default_fields_for_post
```
Comment (ehmadzubair) [inline digital_benchmark/facebook_benchmark/data_provider.py]: `post_fields = fields or FACEBOOK_DEFAULT_FIELDS_FOR_POST`

Comment (ehmadzubair) [reply digital_benchmark/facebook_benchmark/data_provider.py]: or acts as a guard operator

File: digital_benchmark/facebook_benchmark/views.py [hunk 10]
```diff
@@ -1,3 +1,32 @@
 from django.shortcuts import render
+from django.http import HttpResponse
+
+import facebook as fb
+
+version = '3.1'
```
Comment (ehmadzubair) [inline digital_benchmark/facebook_benchmark/views.py]: I believe we're already using it in settings

File: digital_benchmark/facebook_benchmark/views.py [hunk 11]
```diff
@@ -1,3 +1,32 @@
...
 
 # Create your views here.
+def feed(request, access_token):
+    graph = fb.GraphAPI(access_token=access_token, version=version)
```
Comment (ehmadzubair) [inline digital_benchmark/facebook_benchmark/views.py]: Move this to templates

Comment (ehmadzubair) [reply digital_benchmark/facebook_benchmark/views.py]: and make this a class based view rather than a function based view

**Comment by ehmadzubair** [comment]
Also, add a requirements.txt file and populate it from pip freeze

File: digital_benchmark/digital_benchmark/settings.py (hunk 3 above)
Comment (ehmadzubair) [reply digital_benchmark/digital_benchmark/settings.py]: Make them a list in the settings. We can join them where we use them. Pl discuss this with me tomorrow.

File: digital_benchmark/facebook_benchmark/views.py (hunk 11 above)
Comment (sherali-arbisoft) [reply digital_benchmark/facebook_benchmark/views.py]: class-based views or class-based generic views

**Comment by ehmadzubair** [comment]
class based views for now

File: digital_benchmark/facebook_benchmark/models.py [hunk 12]
```diff
@@ -1,3 +1,98 @@
...
+
+class ReactionChoice(Enum):
+    NONE = 'None'
+    LIKE = 'Like'
+    LOVE = 'Love'
+    WOW = 'Wow'
+    HAHA = 'Haha'
+    SAD = 'Sad'
+    ANGRY = 'Angry'
+
```
Comment (ehmadzubair) [inline digital_benchmark/facebook_benchmark/models.py]: add choices.

File: digital_benchmark/facebook_benchmark/models.py [hunk 13]
```diff
@@ -1,3 +1,98 @@
 from django.db import models
+from django.db import models
+
+from enum import Enum
+
+class SoftDeleteMixin(models.Model):
```
Comment (ehmadzubair) [inline digital_benchmark/facebook_benchmark/models.py]: CreateUpdateMixin

File: digital_benchmark/facebook_benchmark/views.py (hunk 11 above)
Comment (sherali-arbisoft) [reply digital_benchmark/facebook_benchmark/views.py]: TODO: move to templates after creating models and storing data using models

**Comment by ehmadzubair** [comment]
@sherali-arbisoft remember one branch per feature

**Pull request merged**